### DISCUSSION TITLE
Aiguillage - personnalisé texte selon poursuite ou URL (cul de sac)

### DISCUSSION THREAD

**Discussion by florentarneodo** [MTESSDev/FRW#125] [Annonces]
Bonjour,

Dans nos formulaires nous utilisons l'aiguillage afin d'éviter que des gens remplissent les formulaires pour rien. J'appelle ça + communément un "cul de sac" dans mon langage si on ne veut pas que le client poursuive.

Exemple fictif : êtes-vous Québécois?
> si oui : poursuivre sa demande vers le formulaire
> si non : on affiche "ce n'est pas pour toi" dans un beau encadré (par exemple). Cependant, il peut toujours faire "suivant" puis "poursuivre ma demande" pour arriver sur un site informationnel (option URL de l'aiguillage) qui lui explique ses options, etc.

Or on ne trouve pas du tout cohérent que la personne voit **"vous pouvez poursuivre votre demande" après lui avoir écrit qu'il ne pouvait pas faire de demande**. Est-il possible de personnaliser ce texte (et le bouton bleu) dans le cas où on choisit l'option URL dans l'aiguillage?

![aiguillage-poursuivre-ma-demande](https://github.com/MTESSDev/FRW/assets/140817361/f8fb721b-9829-480b-95d0-3369336cbb2f)

Merci et bonne journée.

PS : votre exemple est parfait pour illustrer le propos, choisissez de finir sur Google après avoir cliqué sur "poursuivre ma demande":
https://formulaires.it.mtess.gouv.qc.ca/Form/700000/TEST_AIG/0/5lqJ734lt0Orevw7ZVG7bg

**Comment by LarcM7** [comment]
Bonjour,
Il n'est pas possible de personnaliser le texte de ce bouton en fonction de chaque option d'aiguillage.
Il est toutefois possible de personnaliser le texte du bouton, mais ça reste le même texte pour toutes les options.

Quelques options différentes seraient possibles dans votre situation :

**Contrôler l'accès à ce formulaire dans votre système autorisé**

Dans le cas ou votre système connaitrait l'info du citoyen sur laquelle l'éligibilité est déterminée :  vous pourriez choisir d'offrir ou de pas offrir ce formulaire au citoyen.

**Laisser le citoyen soumettre le formulaire** (solution recommandée✔)

Lorsqu'il répond à la question qui vous permet de déterminer qu'il n'est pas éligible, masquer le reste des questions, faisant en sorte que la section suivante pour lui sera la section pour soumettre le formulaire. Ne pas l'informer immédiatement avec un encadré. Au moment de soumettre son formulaire, il arrivera sur la page informationnel que vous avez prévu à cet effet et qui lui expliquera tout ce qu'il a à savoir dans sa situation.

Avantages : 
- évite de laisser "traîner" un formulaire démarré non complété dans l'Espace client de cet utilisateur, qu'il devra supprimer manuellement (ou qui sera épuré après 6 mois)
- laisse une trace qu'un utilisateur a voulu remplir une demande mais qu'il n'était pas éligible (si jamais votre Ministère à besoin d'établir une volumétrie sur cette clientèle)
- probablement plus clair pour l'utilisateur d'arriver dans votre page informationnelle dans votre site que de se faire informer via un encadré qui apparaît dynamiquement

**Masquer la section Révision lorsqu'il n'est pas éligible.**

Bien que cette situation ne soit pas idéale, vous pouvez choisir de masquer la section Révision lorsque l'utilisateur n'est pas éligible. Dans l'encadré qui informe l'utilisateur, vous lui dites que malheureusement il n'est pas éligible, que ça sert à rien de continuer, qu'il n'a qu'à fermer cette fenêtre maintenant (ou naviguer vers votre page informationnelle).

Inconvénients : 
- laisse un formulaire démarré mais non complété associé à cet utilisateur. Il peut ensuite le supprimer lui-même (si vous leur offrez cette possibilité dans votre Espace client) ou il sera épuré 6 mois plus tard
- ne permet pas d'établir une volumétrie
- court-circuite le cycle de vie du formulaire 
- vous devez vous assurer de mettre l'encadré d'information sur chaque question qui peut discriminer son éligibilité 

En espérant que cela vous aide. N'hésitez pas s'il y a quoi que ce soit.

Merci et bonne journée.

**Reply by florentarneodo** [reply]
Bonjour,

Je pense que ça sera plus simple d'en parler durant notre rencontre le lundi PM. L'objectif de l'aiguillage est justement de lui poser des questions dont on n'a pas la réponse, donc ta proposition 1 est écartée.

Proposition 2 : côté expérience client, c'est complètement incohérent de le faire cliquer sur "continuer ma demande" pour l'amener sur un "désolé mais ce n'est pas pour toi"

Proposition 3 : il n'arrive pas au même bouton "poursuivre ma demande" si on enlève la révision? Le but de l'aiguillage est justement de ne pas avoir les inconvénients que tu écris, c'est pourquoi je pense que c'est mieux d'en parler de vive-voix pour être sûr de se comprendre.

Merci et bonne journée.

**Reply by LarcM7** [reply]
La proposition 2 n'est pas aussi incohérente que vous le dites. Le citoyen remplit un formulaire pour un service xyz, c'est tout à fait normal qu'il soumette sa demande et que le Ministère fasse une analyse de l'éligibilité selon les réponses fournies. Malheureusement, dans son cas, il n'est pas éligible il n'y a pas 1000 façons de lui dire. 

S'il remplissait le formulaire papier, il n'y aurait pas un agent qui regarderait par-dessus son épaule au fur et à mesure qu'il remplit pour lui enlever sa copie des mains aussitôt qu'il répond à une question en lui disant "oublie ça finalement t'es pas éligible". La personne soumettrait son formulaire et recevrait sa réponse par la suite.

Ne mettez pas le texte "Continuer ma demande" si c'est ce qui vous dérange à ce point dans cette situation. Pourquoi ne pas utiliser tout simplement "Soumettre"? L'analyse du formulaire est effectuée en quelques secondes et le citoyen à son résultat dans votre site : "_Vous n'êtes pas éligibles, voici les explications, voici vos options, etc._".

"Soumettre" n'est peut-être pas idéal dans le cas de quelqu'un qui est éligible (c'est mieux "Poursuivre ma demande" dans cette situation). Mais vous pouvez personnaliser le texte sur la page révision. Par exemple :  

![image](https://github.com/MTESSDev/FRW/assets/129791924/16874cea-7d97-4d2b-8af1-badf50e09270)

Concernant la proposition 3, non effectivement il ne peut pas "soumettre" son formulaire (ou "Poursuivre sa demande"). 

Je comprends que vous voulez utiliser l'aiguillage, mais que vous êtes inconfortable avec le texte du bouton qui ne peut pas être ajusté en fonction de l'aiguillage. Malheureusement pour l'instant ce n'est pas quelque chose que nous avions prévu de faire. 

Vous pourriez aussi prévoir un texte qui explique à vos utilisateurs, au début à la première section, qu'il aura à soumettre un premier formulaire pour déterminer son éligibilité et que s'il est éligible il pourra remplir le 2e formulaire.

Bref, il y a plusieurs façons de s'en sortir, je pense, malgré le fait qu'il ne soit pas possible malheureusement, pour l'instant, d'ajuster le texte du bouton en fonction de l'aiguillage. 

Vous pourrez en discuter si vous voulez en rencontre un lundi! 

Merci et bonne journée

**Reply by LarcM7** [reply]
Autre idée qui pousse un peu plus loin une des suggestions de mon dernier message : 

Dans la section "Revision", doublez les 2 composants "dynamic" et bonifier le v-if déjà présent avec la condition sur laquelle repose l'éligibilité du citoyen (ex.: Québécois? : non).

De cette façon vous pourriez personnaliser le texte affiché au-dessus du bouton en fonction de l'aiguillage. Ce n'est pas le texte du bouton, mais presque! 😁

**Comment by vinnyboii** [comment]
Bonjour,
Tel que discuté, vous pouvez changer le texte du bouton pour "Poursuivre" afin qu'il s'applique à toutes les situations de votre formulaire.
```
config:
  soumission:
    texteBoutonSoumettre:
      fr: Poursuivre
      en: Continue
```
Et comme Marc l'a mentionné dans sa réponse ci-dessus, vous pouvez doubler le composant dynamic pour modifier le texte affiché au dessus du bouton selon vos conditions. Voici le composant à rechercher dans votre configuration:
```
            - type: dynamic
              tag: div
              v-if: val('EtatRevision') === 'sans-erreur' 
              classes: texte-revision
              text: 
                fr: Votre formulaire est validé. Vous pouvez poursuivre votre demande.
                en: Your form has been validated. You can click on the Submit button.
```

Merci et bonne journée!

**Reply by jodionrq** [reply]
Bonjour,
Dans le cas où on veut doubler le v-if qui afficherait le texte différent, est-ce qu'on se servirait des val comme pour le mode redirection dans la config?
Si je prends l'exemple plus haut "Québécois? : non" ça donnerait 
v-if: val('Quebecois') === 'non'
classes: texte-revision
text:
  fr: Selon vos réponses, vous n'êtes pas admissible pour faire une demande, veuillez cliquer sur le bouton "Poursuivre" pour plus d'explication.

**Comment by j0ber** [comment]
Bonjour,
Pour l'affichage conditionnel de texte c'est effectivement avec un v-if que vous allez le faire. Comme vous le faites déjà à plusieurs endroits dans votre formulaire.
"val" permet d'obtenir la valeur d'un champ ex. v-if: val('NomDeMonChamp') === 'uneValeur'

Il est important de conserver la condition actuelle  sur l'état de révision... Vous devriez simplement ajouter un component dynamic à l'intérieur de celui-ci et y ajouter vos 2 blocs de textes conditionnels. Ça donnerait un truc du genre :
```
- type: dynamic
  tag: div
  v-if: val('EtatRevision') === 'sans-erreur'  
  components:
    - type: dynamic
      tag: div
      v-if: val('VotreChamp1') === 'Votre valeur1'
      classes: texte-revision
      text: 
        fr: Ici votre texte1
        en: Ici votre texte1.
    - type: dynamic
      tag: div
      v-if: val('VotreChamp2') === 'Votre valeur'2
      classes: texte-revision
      text: 
        fr: Ici votre texte2
        en: Ici votre texte2.

```

**Reply by jodionrq** [reply]
Merci pour l'exemple!
Si j'applique ce code pour valider plusieurs champs et faire apparaître un texte qui correspond à une série de réponses, il faudrait que toutes les réponses soient "bonnes" ou "mauvaises". Car si certaines questions sont "bonnes" et d'autres "Mauvaises", les deux messages s'affiche. Est-ce qu'il y a une solution?
![image](https://github.com/MTESSDev/FRW/assets/124822406/e62376a9-a77a-4ec6-93cb-554f3988ea44)

**Comment by j0ber** [comment]
Je ne sais pas ce que vous essayez de faire exactement, mais ce sont vos conditions v-if qui ne sont pas bonnes... Leur combinaison ne fait pas de sens. Vous utilisez des OU ( || ).
Vous devez les revoir... 

C'est peut-être des ET (&&) dont vous avez besoin? FRW n'est pas en problème ici, c'est la logique de vos conditions qui l'est.

**Comment by jodionrq** [comment]
Avec des && il n'y a rien qui s'affiche. À moins qu'on utilise un v-else? Il y a plusieurs questions de validation.

**Reply by j0ber** [reply]
Je ne sais pas ce que vous tentez de faire exactement. Qu'est-ce que vous tentez de vérifier comme condition? La condition actuelle ne fait aucun  sens c'est certain. Pourquoi vous parlez de plusieurs questions de validation alors que votre condition n'en contient que deux, et les 2 mêmes pour chaque v-if?

**Comment by jodionrq** [comment]
Je voulais commencer par tester avec deux questions pour vérifier si mon code fonctionnait bien.
Est-ce que ce serait possible de se parler via Teams, ce serait surement plus simple de mieux vous expliquer mon cas.
Vous pouvez me rejoindre : jonathan.dion@retraitequebec.gouv.qc.ca
Merci!

**Comment by FujuDev** [comment]
**Solution trouvée :**

Usage de **&&** dans le premier _v-if_ afin de s'assurer du seul chemin valide.

Pour le deuxième _v-if_ qui doit donc être vrai lorsque nous avons la condition contraire du premier, il faut copier coller le _v-if_ dont l'on souhaite avoir son _else_, et rajouter ! devant la condition, ainsi que d'entourer cette dernière par des parenthèses.

Le ! étant un caractère clé pour yaml, il est nécessaire de mettre toute notre chaine de caractère de condition entre des guillemets " afin de spécifier que l'on rentre bien une chaine de caractère.

Voici le résultat final :

![image](https://github.com/MTESSDev/FRW/assets/31103239/a386f6bf-6eb3-4f68-8d25-f34b162836ab)
